STM32F467 enabled bootloader

davids5 opened the pull request:
This adds PX4 FMU V4 PRO and PX4 IO V3 bootloaders. 

Currently I am waiting on libopencm3 ( @karlp ) to merge PX4 contributions to their upstream for the USB and RCC of the STM32F469 and STM32F446.
- [x]  Need PX4 contributions upstream-ed to libopencm3 
    - Support for the PLL was incorporated upstream   
    - A work around has been added fir the USB driver issues,
- [ ]  Need to test Safety Switch BL activation and LED Safety Switch indication (waiting on cables from @pkocmoud  - Defered
- [ x] Validate RTC wait for BL 
  - tested with a forced load of V2 FW
- [ ] need PX4/Firmware for PX4 IO V - defered

LorenzMeier commented:
@davids5 Is the BL pinout verified against the v4 reference design?

kd0aij commented:
@davids5 Which version of arm-none-eabi-gcc are you using to compile? I'm getting a flash overflow for px4iov3 using 4.8.

davids5 commented:
@kd0aij  4.9.3 q3 I think

davids5 commented:
@LorenzMeier  I coded to auav semantics not the spread sheet. I will cross check it Mondays

LorenzMeier commented:
It has to follow the spreadsheet to qualify for FMUv4, so its a good gate to check it here.

pkocmoud commented:
@LorenzMeier - Can you suggest a link to the "Reference" spreadsheet as this may be different from my original one.

@davids5 - Please remove the reference to FMUv4 until we have sorted this out.

davids5 commented:
@LorenzMeier RC5 only has 2 signals that deviate

PH2						LIS3_CS	
PH3						LIS3_DRDY	

@pkocmoud says however that there is a RC6 - he will send me shortly - So I need to check that also.

As far as libopencm3 (@karlp) goes, Unless they merge contributions to upstream this week- I will have to stub out that code into this repo.

davids5 commented:
@LorenzMeier, @pkocmoud , @kd0aij 

1) Since I coded this to RC5 it is compatible - RC7 will fix RC6 and make it also compatible.

2) Note I have not changed the remote for libopencm3 in hope that @karlp will get the 2 PR's in.
In the mena time, if you need to build  set the URL to for https://github.com/davids5/libopencm3.git and check out master_rcc_usb

davids5 commented:
@LorenzMeier I have changed the title to  to DO NOT MERGE

To do the F476 bootloader I needed to rev libopencm3 to their latest master that picked up the PLL support for the F467. In regression testing fmuv2's bootloader I see random timeouts on erase. 

```Erase  : [====================] 100.0% (timeout: 0 seconds) 
ERROR: timed out waiting for erase
```
It works the second time after the erase and sometimes 2 times in a row before it fails randomly. Master on the old rev of libopencm3 - is solid every time

I have rebased my 2 bootloader branches on master and the only change here in from the fmuv2 perspective is the libopencm3, mostly naming in the API. 

So I have got some debugging serious debugging to do.

LorenzMeier commented:
I'm happy to provide the funding to move to the NuttX API for the bootloader. I think it's time to consolidate.

davids5 commented:
@LorenzMeier - good news!

It turn out that the root cause of my problem was a hardware issue. With the USB debugger in circuit. The  type B USB cable was dropping enough voltage, it was marginal. 

I can only assume that the updated libopencm3 code is drawing ever slightly more current on erase, then the last master. 

So for now I am unblocked.

davids5 commented:
I am merging this as the px4fmuv4 Pro bootloader is usable as is.
